proof/refine/ARM/Ipc_R.thy

```diff
@@ -5223,26 +5274,268 @@ lemma replyPush_sch_act_wf:
+  apply (rule hoare_seq_ext[OF _ gts_sp'])
+  apply (case_tac state; clarsimp)
+         \<comment> \<open>BlockedOnReceive\<close>
+         apply (wpsimp wp: replyUnlink_replyTCB_None getEndpoint_wp
+                           hoare_vcg_all_lift hoare_vcg_imp_lift')
+         apply (clarsimp simp: pred_tcb_at'_def obj_at'_def projectKOs)
+         apply (drule_tac ko="ko :: reply" for ko in sym_refs_ko_atD'[rotated])
+          apply (fastforce simp: obj_at'_def projectKOs)
+         apply (clarsimp simp: tcb_st_refs_of'_def tcb_bound_refs'_def
+                               refs_of_rev' get_refs_def ko_wp_at'_def
+                        split: option.splits if_splits)
+        \<comment> \<open>BlockedOnReply\<close>
+        apply (wp replyUnlink_replyTCB_None gts_wp' hoare_vcg_all_lift
+                  hoare_vcg_const_imp_lift hoare_vcg_imp_lift'
+               | rule threadSet_pred_tcb_no_state
+               | simp add: replyRemoveTCB_def cleanReply_def if_fun_split)+
+        apply (clarsimp simp: pred_tcb_at'_def obj_at'_def projectKOs)
+        apply (drule_tac p=rptr and ko="ko :: reply" for ko in sym_refs_ko_atD'[rotated])
+         apply (fastforce simp: obj_at'_def projectKOs)
+        apply (clarsimp simp: tcb_st_refs_of'_def tcb_bound_refs'_def
+                              refs_of_rev' get_refs_def ko_wp_at'_def
+                       split: option.splits)
+       \<comment> \<open>Other thread states\<close>
+       apply (all \<open>wpsimp simp: cancelSignal_def sym_refs_asrt_def wp: hoare_drop_imps\<close>)
+       apply (all \<open>clarsimp simp: pred_tcb_at'_def obj_at'_def projectKOs\<close>)
+       apply (all \<open>drule_tac p=rptr and ko="ko :: reply" for ko in sym_refs_ko_atD'[rotated]\<close>)
+             apply (fastforce simp: tcb_bound_refs'_def tcb_st_refs_of'_def refs_of_rev'
+                                    get_refs_def ko_wp_at'_def obj_at'_def projectKOs
+                             split: option.splits)+
+  done
+
+lemma setThreadState_BlockedOnReply_invs':
+  "\<lbrace>\<lambda>s. invs' s \<and> tcb_at' t s \<and> ep_at' eptr s \<and> ex_nonz_cap_to' t s \<and>
+        valid_bound_reply' rptr s \<and> sch_act_not t s \<and> t \<noteq> ksIdleThread s\<rbrace>
+   setThreadState (BlockedOnReceive eptr cg rptr) t
    \<lbrace>\<lambda>rv. invs'\<rbrace>"
   apply (simp add: invs'_def valid_state'_def)
-  apply (wp setThreadState_ct_not_inQ)
-  apply (fastforce dest: global'_no_ex_cap)
+  apply (wp sts_sch_act' setThreadState_ct_not_inQ valid_irq_node_lift)
+  apply (clarsimp dest: global'_no_ex_cap simp: valid_tcb_state'_def comp_def)
+  done
+
+crunches cancelIPC
+  for sch_act_not[wp]: "sch_act_not t"
+  and ep_at'[wp]: "\<lambda>s. P (ep_at' eptr s)"
+  and valid_bound_reply'[wp]: "valid_bound_reply' reply"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma setReply_valid_ep'[wp]:
+  "setReply rptr r \<lbrace>valid_ep' ep\<rbrace>"
+  unfolding setReply_def valid_ep'_def
```

Comment:
This one, too. Would `set_reply'.typ_at_sc_at’_n_lifts’` work instead? With some simplification?